### DISCUSSION TITLE
group packages by package manager

### DISCUSSION THREAD

**Discussion by electriquo** [renovatebot/renovate#16419] [Request Help] [answered]
Renovate supports many [package mangers](https://docs.renovatebot.com/modules/manager/) and [grouping](https://docs.renovatebot.com/presets-group/).

In a project with multiple package managers, how do I configure renovate to group all related updates by package manager into the same pull-request?

For instance, open a single pull-request for npm package manager; and do the same for each package manager.

**Comment by rarkins** [comment]
You could try this:

```
  "groupName": "{{manager}}"
```

**Reply by electriquo** [reply]
It does not work for me or I did something wrong. My configuration is

```json
{
  "extends": ["config:base", "group:recommended"],
  "groupName": "{{manager}}",
}
```

but Renovate group all changes into 1 pull-request and not per manager.
What am I doing wrong? (I tried to remove `group:recommended` and re-run Renovate through the [Dependency Dashboard](https://docs.renovatebot.com/key-concepts/dashboard/) but to no avail).

To be clear, when I look at the [Dependency Dashboard](https://docs.renovatebot.com/key-concepts/dashboard/), It lists the detected dependencies of my project:
- docker-compose
- dockerfile
- github-actions
- npm
- nvm
- regex

and I would like Renovate to group pull-requests per manager (or detected dependencies). For example, if there are many npm packages that have been updated, then group them into the same pull-request rather than having multiple pull-request for each npm package.

Also tried to use [matchManagers](https://docs.renovatebot.com/configuration-options/#matchmanagers) in packageRules, but matchManager does not accept a wildcard same as [matchPackagePatterns](https://docs.renovatebot.com/configuration-options/#matchpackagepatterns).

**Comment by PhilipAbed** [comment, marked as answer]
try something like this maybe 
```
"packageRules": [
    {
      "matchManagers": ["docker-compose", "npm", "nvm","github-actions","docker-file","regex"],
      "groupName": "{{manager}}"
    }
]
```

**Reply by electriquo** [reply]
Will try that. Is there a way to use something more generic for the matchManagers (a wildcard is not accepted) rather than specifying explicitly each manager?

**Reply by PhilipAbed** [reply]
nope it accepts an array of strings not regex,
usually people group dependencies by Datasource not package managers, 
but that also is an array of strings
the number of datasources/package managers are few, so it doesn't make sense to use a wild card

**Reply by electriquo** [reply]
Using a wildcard would be really handy. It frees one from knowing which package managers are expected by matchManagers and update the array per package managers detected in a project by Renovate.

**Reply by rarkins** [reply]
If you want this to apply to all managers then `"matchPackagePatterns": ["*"]` would also work and then you con't need `matchManagers`

**Reply by sheldonhull** [reply]
I noticed when doing this the gomod, npm, github action, etc all showed up as "major regex" which was odd. I thought it would group gomod with go etc.

**Comment by electriquo** [comment]
Do you think we should add "group:byManager" preset to [presets group](https://docs.renovatebot.com/presets-group/)?

**Reply by viceice** [reply]
no, i don't think it's widely used

**Reply by electriquo** [reply]
it will once we it will be added :)

**Reply by fregante** [reply]
I think this makes perfect sense though. If one is using multiple languages, `group:all` doesn't make sense, but `group:byManager` would. If anything, it should be suggested instead of `group:all`

**Reply by rarkins** [reply]
PR welcome then

**Comment by stempler** [comment]
Maybe useful for someone else, a configuration to only group minor and patch updates by manager:

```
  "packageRules": [
    {
      "groupName": "{{manager}} non-major dependencies",
      "groupSlug": "{{manager}}-minor-patch",
      "matchPackagePatterns": [
        "*"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ]
    }
  ],
```

For me solved the issue that the `group:allNonMajor` preset sometimes leads to invalid PRs when `gradle-wrapper` manager updates were mixed with other updates.

**Comment by acestronautical** [comment]
Just want plus one that  `group:byManager` would be nice to have as a preset:
```
{
  "groupName": "{{manager}} dependencies",
  "groupSlug": "{{manager}}",
  "packageRules": [
    {
      "groupName": "{{manager}} dependencies",
      "groupSlug": "{{manager}}",
      "matchPackagePatterns": [
        "*"
      ]
    }
  ],
  "separateMajorMinor": false
}
  ```
